src/server.c

```diff
@@ -133,23 +154,47 @@ void serverLogRaw(int level, const char *msg) {
     } else {
         int off;
         struct timeval tv;
-        int role_char;
         pid_t pid = getpid();
         int daylight_active = atomic_load_explicit(&server.daylight_active, memory_order_relaxed);
 
         gettimeofday(&tv, NULL);
         struct tm tm;
         nolocks_localtime(&tm, tv.tv_sec, server.timezone, daylight_active);
-        off = strftime(buf, sizeof(buf), "%d %b %Y %H:%M:%S.", &tm);
-        snprintf(buf + off, sizeof(buf) - off, "%03d", (int)tv.tv_usec / 1000);
+        switch (server.log_timestamp_format) {
+        case LOG_TIMESTAMP_LEGACY:
+            off = strftime(buf, sizeof(buf), "%d %b %Y %H:%M:%S.", &tm);
+            snprintf(buf + off, sizeof(buf) - off, "%03d", (int)tv.tv_usec / 1000);
+            break;
+
+        case LOG_TIMESTAMP_ISO8601:
+            off = strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S.", &tm);
+            char tzbuf[7];
+            formatTimezone(tzbuf, sizeof(tzbuf), server.timezone, server.daylight_active);
+            snprintf(buf + off, sizeof(buf) - off, "%03d%s", (int)tv.tv_usec / 1000, tzbuf);
+            break;
+
+        case LOG_TIMESTAMP_MILLISECONDS:
+            snprintf(buf, sizeof(buf), "%lld", (long long)tv.tv_sec * 1000 + (long long)tv.tv_usec / 1000);
+            break;
+        }
+        int role_index;
         if (server.sentinel_mode) {
-            role_char = 'X'; /* Sentinel. */
+            role_index = 0; /* Sentinel. */
         } else if (pid != server.pid) {
-            role_char = 'C'; /* RDB / AOF writing child. */
+            role_index = 1; /* RDB / AOF writing child. */
         } else {
-            role_char = (server.primary_host ? 'S' : 'M'); /* replica or Primary. */
+            role_index = (server.primary_host ? 2 : 3); /* Slave or Master. */
+        }
+        switch (server.log_format) {
+        case LOG_FORMAT_LOGFMT:
+            fprintf(fp, "pid=%d role=%s timestamp=\"%s\" level=%s message=\"%s\"\n", (int)getpid(), roles[role_index],
+                    buf, verbose_level[level], msg);
```

Comment:
I think this PR looks good now, almost ready to merge.

I'm just thinking about the corner case that log messages containing special characters like `"`, backslash or linebreaks. I believe modules can log any text they want. (Is that true? I didn't check.) I couldn't find if logfmt supports escaping. Maybe we can just replace some forbidden chars with other chars or something.

Ideally we should have a test case with a module that tries to log something like that. Maybe it's difficult. Do we have anything in our test framework to check what we to get in the logs?

If we add `LOG_FORMAT_JSON` in the future, it will need a similar but maybe slightly different check that it's correct UTF-8 and maybe do some escaping of characters like backslash.

Reply:
Not sure if that's a question for me. 

Modules are logging messages through this routine currently, and following the c string rule, such as backslash+n means newline? Why do we need new rules for logfmt?
One reason I can see is that double quotes is too commonly used in users messages. We can define something more complicated thing to brace message and timestamp. Such as !?+message+?!
Plus, anything we do to the message itslef will not be transparent to users. And they'll have to change their message parser.

Reply:
It's a question for you and others. :)

> Modules are logging messages through this routine currently, and following the c string rule, such as backslash+n means newline? Why do we need new rules for logfmt?

Modules can log stuff using the function below, which is just a wrapper for serverLog. It's implemented as `VM_Log` in `module.c`.

              void ValkeyModule_Log(ValkeyModuleCtx *ctx,
                                    const char *levelstr,
                                    const char *fmt,
                                    ...);

Let's say a module logs something that contains a key name or something that we can't control. It would make it possible to mess up the log if we don't do anything special to the message. Let's say the message content is the following two-lines string:

    x"
    pid=9999 timestamp=never level=banana message="y

If we just include this as is, it will appear as message "x" and then an injected log line with dummy information and message "y" logged below the first line. (If the module is written in C, then line breaks can be written as `\n` in string literals, but it can also come from e.g. a key name or a command argument that comes from a client and is combined using `snprintf` or in some other way.)

> One reason I can see is that double quotes is too commonly used in users messages. We can define something more complicated to brace message and timestamp. Such as !?+message+?!

Is that possible in logfmt? I haven't found a formal syntax of logfmt but I would guess that all logfmt parsers expect double quotes `"message"`?

> Plus, anything we do to the message content will not be transparent to users. And they'll have to change their message parser.

Yes, that can be a problem. The log message created from within Valkey can be designed to avoid using `"`. For log messages from modules, it is probably fine to forbid it (change `"` to `'` and remove all line breaks) just to make sure the log isn't broken.

Reply:
Code updated as your suggestion. 
Please note that I've modified the mutability of `*msg`.